Added a Publish() and Subscribe() method onto the Hub class to allow …

codearoo opened the pull request:
Hi.. I love the simplicity of this project. However I did find a significant bug and need for small change due to the use of the static instance of a Hub in the Extensions class. While it is very handy to use a static instance, in a multi-threaded web environment it typically causes dangers and users might want to limit an instance of a Hub to their current thread. I added the 2 methods to the Hub class to make it possible to directly use that hub.

I also added some test code to demonstrate the current extension methods for publish which weren't tested at all, as well as the direct use which actually did not work as someone would expect without this fix as far as I can tell.

codearoo commented:
Actually I realized that I miss-used the 'sender' as 'this' because in the Publish it wasn't doing anything with that. So I'm refactoring a bit but the point is to allow use of the Hub directly and also to test all the Extensions as it is expected for them to work.

The other thing is that the Unsubscribe has the same issue, so I'm addressing that as well.

codearoo commented:
If someone is directly publishing, subscribing and unsubscribing to the Hub, because of the way the Object Extensions are written, it makes the illusion that one could do this for a specific hub they are accessing. With the new methods I added, this becomes the reality. The difference though, is that the "sender" ends up being the Hub itself. I'm thinking that is totally fine. The only purpose I can see for another object to be a 'sender' is that if that object disappears, the Hub will not act on those actions, however that is mostly a cleanup due to somebody forgetting to properly Unsubscribe. So while these new methods change who the 'sender' is, the functionality is closer to what I think the user would expect and it has no negative implications.

upta commented:
I'll definitely take a look over it once you've made your refactors, but I'm a bit confused.  In what way are you unable to new up an instance of Hub yourself and interact with it, just like the extensions do via their static instance?  They're really just there as a thin, convenience mechanism on top of Hub and not required if you feel like having your own Hub object.

Yeah, I just noticed that sender bit in Publish (I honestly have no idea why that was ever passed in, looks like it was never used.

I'm also quite curious about your use-case in a web app (I honestly never even considered using it in one, not that you should be unable to do so.  I originally built this for use in a Xamarin Android app).

Thanks!

codearoo commented:
Well I added test code that tests the use of the extensions which was missing. And then I added test code using a hub instance using those same methods that would easily fool a user developer. Because if one sees that they can do               *myhub.Publish(new Event());*   they would *think* they are publishing to THAT hub, but they would be wrong. So the new methods make that true.

The alternative is to throw an exception in those methods telling them it's not allowed, but there's no reason it shouldn't be.

Regarding your question about web app..  that's just one idea.. but it can be used by ANYTHING. The point is that anything multi-threaded would have unexpected issues. A web app might have code to have a Logger subscribe to "Log Events" or some Alerter to subscribe to them or other things, who knows.. anything at all. But if the user starts coding the subscription of those that gets run in each thread, then it will snowball into multiple threads subscribing to the same Hub. So either the developer will make those singletons, or they want a a hub tied just to a web thread or business process.

codearoo commented:
And to clarify, I will NOT be refactoring because of what I wrote after my previous commit regarding the Hub being a sender as being OK.

codearoo commented:
I'd almost say that maybe there is no point in having a 'sender' at all.

upta commented:
I'm with ya, now :)  Basically just using the Hub instance itself as the default "key" for subscriptions, I'm on board with that.

Sender should be removed from Publish, I agree (since, you know, it doesn't actually do anything ;).

As far as the subscription is concerned, it basically just serves as a "key" for cleaning up.  This enables you to unsubscribe from a "set" of handlers all associated by that sender object.  It also enables the automatic cleanup if that object has been disposed.

codearoo commented:
Ya.. can leave the 'sender' in.. it does help for an object to say "unsubscribe me from everything" easily. But the auto cleanup is something that a developer should not rely on.. but should always Unsubscribe before losing the object.

codearoo commented:
I will leave the removal of 'sender' from the Publish to you, unless you want me to do that.

codearoo commented:
@upta  Hi.. just wondering if you're going to merge this branch in or not.

upta commented:
Totally forgot I never merged this, sorry!  Thanks for the reminder

upta commented:
Updated nuget as 1.5.0
